### PR TITLE
Rotation Controls

### PR DESCRIPTION
**Pull request by arjxn-py**
Shall supersede #574

### CODE REVIEW

File: packages/base/src/3dview/mainview.tsx [hunk 1]
```diff
@@ -724,18 +740,21 @@ export class MainView extends React.Component<IProps, IStates> {
...
-            this._clipPlaneTransformControls.setMode('rotate');
-          }
-          break;
+      if (event.key === 'r' && this._clipSettings.enabled) {
+        event.preventDefault();
+        event.stopPropagation();
+        toggleMode(this._clipPlaneTransformControls);
+      }
+
+      if (event.key === 't' && this._transformControls.enabled) {
```
Comment (martinRenou) [inline packages/base/src/3dview/mainview.tsx:754]:
```suggestion
      if (event.key === 'r' && this._transformControls.enabled) {
```

**Review by martinRenou — COMMENTED** [review]
Let's make it 'r' too

**Comment by trungleduc** [comment]
I get this error


https://github.com/user-attachments/assets/c7c405ac-fea0-4b3d-9703-ce1fdd2df1aa

**Comment by martinRenou** [comment]
> I get this error

Also had that one once, but not after. We should check the value set in the shared model (possible NaN?)

File: packages/base/src/3dview/mainview.tsx [hunk 2]
```diff
@@ -423,6 +423,20 @@ export class MainView extends React.Component<IProps, IStates> {
 
         const updatedPosition = new THREE.Vector3();
         updatedObject.getWorldPosition(updatedPosition);
+        const updatedQuaternion = new THREE.Quaternion();
+        updatedObject.getWorldQuaternion(updatedQuaternion);
+
+        const s = Math.sqrt(1 - updatedQuaternion.w * updatedQuaternion.w);
+        const updatedRotation = [
+          [
+            parseFloat((updatedQuaternion.x / s).toFixed(10)),
```
Comment (martinRenou) [inline packages/base/src/3dview/mainview.tsx:432]: I suspect the error comes from `parseFloat` which returns may `NaN`. Do we really need `parseFloat` here? Why not just `updatedQuaternion.x / s`?

Comment (martinRenou) [reply packages/base/src/3dview/mainview.tsx:432]: Probably a division by `0` when s is `0`, there should be a check for this

Comment (arjxn-py) [reply packages/base/src/3dview/mainview.tsx]: > Do we really need `parseFloat` here? Why not just `updatedQuaternion.x / s`?

Yes, it is also sufficient.

> Probably a division by 0 when s is 0, there should be a check for this

Added a fallback condition, thanks :)

**Comment by arjxn-py** [comment]
> I get this error

the NaN shall be fixed by [`9387bcf`](https://github.com/jupytercad/JupyterCAD/pull/603/commits/9387bcf70b5997f5f4ecacd13b7996a84ecee6cd)

**Comment by trungleduc** [comment]
Thanks! it works

**Review by trungleduc — APPROVED** [review]

**Comment by martinRenou** [comment]
Yay!

**Review by martinRenou — APPROVED** [review]

**Pull request merged**